zabbix_host, interface.port parameter is a string

adrianlzt opened the pull request:
##### SUMMARY
interface.port should be an string to be able to use macros in that
value.
This fixes the case when interface.port is a macro (eg.: "{$MACRO}" and
force=false.

Because, until now, setting the interface.port to an integer was the
correct way to work with force=false, a type validation has been added
to that parameter.
Previously, if an string was used for interface.port, it was converted
to an integer, comparasion didn't work (if interface not in interfaces)
and the module tried to register the same interface twice, returning an
error.

Zabbix API manual specifies that only 'main, 'type', 'useip' and 'bulk'
are integers.
https://www.zabbix.com/documentation/current/manual/api/reference/hostinterface/object

Fixes #63449 and PR 63637
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
zabbix_host

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

Executing this playbook twice will show the error fixed in this PR:
```
- hosts: localhost
  gather_facts: false
  tasks:
    - name: "Register cluster virtual host n Zabbix"
      zabbix_host:
        force: "false"
        server_url: "http://zabbix"
        login_user: "Admin"
        login_password: "zabbix"
        host_name: "somehost"
        description: "Added by Ansible"
        host_groups: "somegroup"
        inventory_mode: "manual"
        interfaces:
          - type: 1
            main: 1
            useip: 0
            ip: 127.0.0.1
            dns: "somehost"
            port: "{$MACRO}"
```

adrianlzt committed:
zabbix_host, interface.port parameter is a string

interface.port should be an string to be able to use macros in that
value.
This fixes the case when interface.port is a macro (eg.: "{$MACRO}" and
force=false.

Because, until now, setting the interface.port to an integer was the
correct way to work with force=false, a type validation has been added
to that parameter.
Previously, if an string was used for interface.port, it was converted
to an integer, comparasion didn't work (if interface not in interfaces)
and the module tried to register the same interface twice, returning an
error.

Zabbix API manual specifies that only 'main, 'type', 'useip' and 'bulk'
are integers.
https://www.zabbix.com/documentation/current/manual/api/reference/hostinterface/object

Tests are changed to use always str in the interface.port.

Two new tests are added.
First one is to show that now registering a host with force=false and
a macro in interface.port works.
The other one tests that interfaces defined using string for port are
compared correctly when force=false is used. Previously it was
a comparasion between an int and a str, interfaces were seen as diferent
and an error was thrown because we were trying to create twice a main
interface.

D3DeFi commented on lib/ansible/modules/monitoring/zabbix/zabbix_host.py:
```diff
@@ -780,6 +780,8 @@ def main():
             if 'main' not in interface:
                 interface['main'] = 0
+            if 'port' in interface and not isinstance(interface['port'], str):
+                module.fail_json(msg="port should be an string on interface '%s'." % interface)
```
I wouldn't go as far as failing module because someone has not provided integer in quotes. This would require us to deprecate the feature first and then remove it in usually 4th release from now (that would be ansible 2.14)

What about this instead?

```suggestion
                try:
                    interface['port'] = str(interface['port'])
                except ValueError:
                    module.fail_json(msg="port should be convertable to string on interface '%s'." % interface)
```

adrianlzt replied:
Way better!

D3DeFi commented:
Thanks for the contribution @adrianlzt ! Please, see my comment above

adrianlzt committed:
Try to kindly convert interface port to string

As suggested, the previous behaviour could break current configurations. This solution accepts integers and strings.

Co-Authored-By: Dusan Matejka <D3DeFi@users.noreply.github.com>

sky-joker commented:
Thank you @adrianlzt for this patch :)

D3DeFi commented:
Looks good, failed CI can be fixed by closing & reopening this PR. Once it is all green I will vote for merge.

D3DeFi commented:
@adrianlzt please close this PR and then click on reopen. CI will rerun and we will be able to vote shipit

D3DeFi reviewed (approved):
Good job!
shipit

D3DeFi commented:
@sky-joker please cast your shipit vote :)

sky-joker commented:
shipit

D3DeFi commented:
bot_status
